documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,486 @@
+- Would require an expensive `join` at the time of data refresh
+- The `image_popularity` view still needs to be refreshed, and will still take
+  an untenably long time.
+- When the data refresh runs, records that have not yet been refreshed in the
+  `image_popularity` view will make it to the API, but without any standardized
+  popularity score.
+
+## Accessibility
+
+<!-- Are there specific accessibility concerns relevant to this plan? Do you expect new UI elements that would need particular care to ensure they're implemented in an accessible way? Consider also low-spec device and slow internet accessibility, if relevant. -->
+
+As part of this project, we should add documentation for the new popularity
+refresh DAGs and the calculations themselves.
+
+## Parallelizable streams
+
+<!-- What, if any, work within this plan can be parallelized? -->
+
+The order of operations for this project is very important.
+
+1. Add `standardized_popularity` column to media tables.
+2. Update the Media stores to accept `standardized_popularity`, defaulted to
+   None, and write this column to tsv.
+3. Update the provider DAGs to calculate `standardized_popularity` at ingestion,
+   and write to the database.
+4. Create the `popularity_refresh` DAG factory.
+5. Enable the `popularity_refresh` DAGs in production, to backfill scores on the
+   media tables.
+6. Once the `popularity_refresh` DAGs have completed their initial run, update
+   the ingestion server to begin copying data from the media tables rather than
+   the matviews.
+7. Remove the popularity steps from the data refresh DAGs.
+8. Drop the materialized views.
+
+Steps 3 and 4 may be done in parallel.
+
+## Rollback
+
+<!-- How do we roll back this solution in the event of failure? Are there any steps that can not easily be rolled back? -->
+
+These changes will only go into effect when we update the ingestion server to
+begin reading from the media tables. We can rollback by reverting just this last
+change. Consequently we should wait to drop the materialized views until we are
+confident that the process is working.
+
+Because we have a far smaller number of `audio` records, we should test the
+process with `audio` first. We should also test the data refresh in staging.
+
+## Risks
+
+<!-- What risks are we taking with this solution? Are there risks that once taken can’t be undone?-->
+
+This approach would add an additional column to our media tables; it also
+requires a temp table used for storing the ids of rows that need to be updated.
+There will be a net decrease in storage needed when the materialized views are
+finally dropped, but in the interim we will use extra space.
+
+We do not have a clear estimate for how long the popularity refresh DAG will
+take, and a complete run is necessary before we can begin running the data
+refresh. If the popularity refresh takes an extremely long time, this will delay
+our ability to restart the image data refresh. _After_ the initial run, we will
```

Comment:
I made a note about this above, but if we potentially performed a one-time copy of existing data from the matview into the new `standardized_popularity` columns on the primary tables, we might be able to unblock a few steps without affecting production data!